File: leetcode/02-two-numbers.ts
// Definition for singly-linked list.
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function (l1, l2) {
  let root = new ListNode(0, undefined);
  let current = root;
  let carry = 0;

  while (l1 || l2 || carry) {
    const sum = (l1?.val ?? 0) + (l2?.val ?? 0) + carry;
    carry = Math.floor(sum / 10);

    current.next = new ListNode(sum % 10, undefined);
    current = current.next;

    l1 = l1?.next ? l1.next : null;
    l2 = l2?.next ? l2.next : null;
  }

  return root.next;
};

File: leetcode/02-two-numbers.py


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        root = ListNode()
        current = root
        carry: int = 0

        while l1 is not None or l2 is not None or carry != 0:
            l1val = 0
            if l1 is not None:
                l1val = l1.val
                l1 = l1.next

            l2val = 0
            if l2 is not None:
                l2val = l2.val
                l2 = l2.next

            current.val = l1val + l2val + carry
            current.next = ListNode()
            current = current.next

        return root
